### ISSUE TITLE
How to extract GIF frames and coalesce them?

### ISSUE THREAD

**Issue by TheShihan** [dlemstra/Magick.NET#420]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am using the latest version of Magick.NET

### System Configuration
- UWP, Min. Version 16229, Target 17763
- Magick.NET version: Magick.NET-Q8-AnyCPU 7.11.1
- Environment: Windows 10 1809 (17763.379), Visual Studio Community 2017

### Question
I try to extract frames from GIF files to show them inside an image control. I tried to do this without Magick.Net and noticed that for some images the frames only contained the deltas.

I saw that Magick.Net has a Coalesce method that can be applied on a MagickImageCollection and if I understand correctly (I am not sure about that) this method should then merge the frames (key frames and delta) into a single (gif) image.

So what I tried to do is to pass frames 0-x to MagickImageCollection to run the Coalesce operation over it but what happens is that I get an image that looks like the first frame.

I created this method to extract the frames:
```
        public static MemoryStream ReadGif(Stream inputStream, int? frameIndex)
        {
            var readSettings = new MagickReadSettings
            {
                FrameIndex = frameIndex
            };
            // Make sure stream is at the beginning
            inputStream.Position = 0;
            var image = new MagickImage(inputStream, readSettings);

            // Save the result
            var memoryStream = new MemoryStream();
            image.Write(memoryStream);
            return memoryStream;
        }
```
My understanding was that by passing the FrameIndex (e.g. x = 9) it would give me a stream of the gif with only frames 0 - 9. So I could then run Coalesce over that. But as told, it gives me back only the first frame.

I then tried to add "FrameCount" to the ReadSettings, with FrameIndex always at 0 and FrameCount for example 9. But I got an exception the I cannot use a FrameCount other than "1" for a Single image (but I do know the GIF has more than 1 frames).

What am I doing wrong or how can GIF frames extracted so that they don't contain only the delta data?

Any help is greatly appreciated.

**Comment by dlemstra** [comment]
You cannot read the individual images when you want the coalesced images from a GIF. This won't work because if you for example need to read image 2 you will need the information from frame 0-2. Image 2 is frame 0 plus the changes from frame 1 and then the changes from frame 2. You will need to do something like this:

```C#
public static MemoryStream ReadGif(Stream inputStream, int frameIndex)
{
    var settings = new MagickReadSettings()
    {
        FrameIndex = 0,
        FrameCount = frameIndex + 1,
    };

    inputStream.Position = 0;
    using (var images = new MagickImageCollection(inputStream, settings))
    {
        images.Coalesce();

        var memoryStream = new MemoryStream();
        images[frameIndex].Write(memoryStream);
        return memoryStream;
    }
}
```

The example above is optimized for the implementation of your example method. If you need all the frames it would probably be better to not supply the `settings` to read all the frames. And then do something with each individual image from the `images` collection.

**Comment by TheShihan** [comment]
Thanks, I got it working with your help.

By the way, to show the resulting image in an UWP Image control I convert it to a SoftwareBitmapSource using this approach:
                    var frameMemoryStream = Utils.ReadGifAndCoalesce(_gifStream.AsStreamForRead(), frameIndex);
                    var decoderForFrame = await BitmapDecoder.CreateAsync(frameMemoryStream.AsRandomAccessStream());
                    var softwareBitmapForFrame = await decoderForFrame.GetSoftwareBitmapAsync();
                    var convertForFrame = SoftwareBitmap.Convert(softwareBitmapForFrame, BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied);
                    var softwareBitmapSourceForFrame = new SoftwareBitmapSource();
                    await softwareBitmapSourceForFrame.SetBitmapAsync(convertForFrame);

The control is then bound to the resulting softwareBitmapSourceForFrame (through a property).

Does this look "sane" to you or maybe is there something "Magic" im Magick.Net to perform this in one line of code? :-)

**Comment by dlemstra** [comment]
I don't know the API of the `SoftwareBitmapSource` class but it would be better to not decode, encode and decode the image. There is not build in `SoftwareBitmapSource` implementation but there is an `ToBitmapSource` method on the `MagickImage` class. This might give you some ideas to just copy the pixels (if that is possible): https://github.com/dlemstra/Magick.NET/blob/3b889b568ba8080287912f9963494b73721a5d82/Source/Magick.NET/Framework/MagickImage.cs#L186?

**Comment by TheShihan** [comment]
Well I had a look at the moment, but to be honest I don't have too much knowledge about those image/pixel data internals, so I don't understand match of what the method does. As it is from .NET 2.0 I tried to get it working with UWP by replacing some classes with the ones available in UWP but not all were.. so I tried a different approach. As the ImageControl can be passed an ImageSource I created following code to transform the stream into an ImageSource that I can then use to show it in an ImageControl:
```
var frameStream = Utils.ReadGifAndCoalesceToStream(_gifStream.AsStreamForRead(), (int)ii);
var bitmapImage = new BitmapImage();
frameStream.Position = 0;
var accessStream = frameStream.AsRandomAccessStream();
await bitmapImage.SetSourceAsync(accessStream);
_gifFrameCollection.Add(new GifFrameInformation { FrameIndex = ii, FrameImageSource = bitmapImage, FrameStream = frameStream.AsRandomAccessStream() }); 
```

This does work. It didn't work when I had not "frameStream.Position = 0;" in it, apparently I have to reset the stream to the beginning.
